## Libraries

In [218]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Analyse exploratoire

In [219]:
df = pd.read_csv("vehicles.csv")

In [220]:
# Dimension du df c-a-d le nombre de lignes et colonnes
df.shape

(426880, 26)

In [221]:
# Nombre et noms de colonnes dans df
df.columns

Index(['id', 'url', 'region', 'region_url', 'price', 'year', 'manufacturer',
       'model', 'condition', 'cylinders', 'fuel', 'odometer', 'title_status',
       'transmission', 'VIN', 'drive', 'size', 'type', 'paint_color',
       'image_url', 'description', 'county', 'state', 'lat', 'long',
       'posting_date'],
      dtype='object')

In [222]:
# Afficher les cinq premières lignes du df
df.head()

id                                                url  \
0  7222695916  https://prescott.craigslist.org/cto/d/prescott...   
1  7218891961  https://fayar.craigslist.org/ctd/d/bentonville...   
2  7221797935  https://keys.craigslist.org/cto/d/summerland-k...   
3  7222270760  https://worcester.craigslist.org/cto/d/west-br...   
4  7210384030  https://greensboro.craigslist.org/cto/d/trinit...   

                   region                         region_url  price  year  \
0                prescott    https://prescott.craigslist.org   6000   NaN   
1            fayetteville       https://fayar.craigslist.org  11900   NaN   
2            florida keys        https://keys.craigslist.org  21000   NaN   
3  worcester / central MA   https://worcester.craigslist.org   1500   NaN   
4              greensboro  https://greensboro.craigslist.org   4900   NaN   

  manufacturer model condition cylinders  ... size  type paint_color  \
0          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
1          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
2          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
3          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
4          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   

  image_url description county state lat long posting_date  
0       NaN         NaN    NaN    az NaN  NaN          NaN  
1       NaN         NaN    NaN    ar NaN  NaN          NaN  
2       NaN         NaN    NaN    fl NaN  NaN          NaN  
3       NaN         NaN    NaN    ma NaN  NaN          NaN  
4       NaN         NaN    NaN    nc NaN  NaN          NaN  

[5 rows x 26 columns]

In [223]:
# Afficher les cinq dernières lignes du df
df.tail()

id                                                url  \
426875  7301591192  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426876  7301591187  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426877  7301591147  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426878  7301591140  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426879  7301591129  https://wyoming.craigslist.org/ctd/d/atlanta-2...   

         region                      region_url  price    year manufacturer  \
426875  wyoming  https://wyoming.craigslist.org  23590  2019.0       nissan   
426876  wyoming  https://wyoming.craigslist.org  30590  2020.0        volvo   
426877  wyoming  https://wyoming.craigslist.org  34990  2020.0     cadillac   
426878  wyoming  https://wyoming.craigslist.org  28990  2018.0        lexus   
426879  wyoming  https://wyoming.craigslist.org  30590  2019.0          bmw   

                           model condition    cylinders  ... size       type  \
426875         maxima s sedan 4d      good  6 cylinders  ...  NaN      sedan   
426876  s60 t5 momentum sedan 4d      good          NaN  ...  NaN      sedan   
426877          xt4 sport suv 4d      good          NaN  ...  NaN  hatchback   
426878           es 350 sedan 4d      good  6 cylinders  ...  NaN      sedan   
426879  4 series 430i gran coupe      good          NaN  ...  NaN      coupe   

       paint_color                                          image_url  \
426875         NaN  https://images.craigslist.org/00o0o_iiraFnHg8q...   
426876         red  https://images.craigslist.org/00x0x_15sbgnxCIS...   
426877       white  https://images.craigslist.org/00L0L_farM7bxnxR...   
426878      silver  https://images.craigslist.org/00z0z_bKnIVGLkDT...   
426879         NaN  https://images.craigslist.org/00Y0Y_lEUocjyRxa...   

                                              description county state  \
426875  Carvana is the safer way to buy a car During t...    NaN    wy   
426876  Carvana is the safer way to buy a car During t...    NaN    wy   
426877  Carvana is the safer way to buy a car During t...    NaN    wy   
426878  Carvana is the safer way to buy a car During t...    NaN    wy   
426879  Carvana is the safer way to buy a car During t...    NaN    wy   

              lat       long              posting_date  
426875  33.786500 -84.445400  2021-04-04T03:21:31-0600  
426876  33.786500 -84.445400  2021-04-04T03:21:29-0600  
426877  33.779214 -84.411811  2021-04-04T03:21:17-0600  
426878  33.786500 -84.445400  2021-04-04T03:21:11-0600  
426879  33.779214 -84.411811  2021-04-04T03:21:07-0600  

[5 rows x 26 columns]

In [224]:
# Affciher les lignes du df, aléatoirement
df.sample(n=10)

id                                                url  \
159959  7309283901  https://masoncity.craigslist.org/ctd/d/center-...   
122040  7316228917  https://atlanta.craigslist.org/nat/cto/d/dulut...   
171067  7303985270  https://huntington.craigslist.org/ctd/d/goshen...   
341640  7316975817  https://providence.craigslist.org/ctd/d/attleb...   
413583  7304249244  https://yakima.craigslist.org/ctd/d/portland-2...   
139495  7311827918  https://bn.craigslist.org/ctd/d/bloomington-20...   
265083  7306975545  https://albany.craigslist.org/ctd/d/cohoes-201...   
156774  7315006117  https://desmoines.craigslist.org/cto/d/des-moi...   
98204   7315809919  https://jacksonville.craigslist.org/cto/d/jack...   
184693  7304988231  https://frederick.craigslist.org/cto/d/frederi...   

                    region                           region_url  price  \
159959          mason city     https://masoncity.craigslist.org  11900   
122040             atlanta       https://atlanta.craigslist.org  12900   
171067  huntington-ashland    https://huntington.craigslist.org  19995   
341640        rhode island    https://providence.craigslist.org  13300   
413583              yakima        https://yakima.craigslist.org      0   
139495  bloomington-normal            https://bn.craigslist.org  17590   
265083              albany        https://albany.craigslist.org  29900   
156774          des moines     https://desmoines.craigslist.org   3500   
98204         jacksonville  https://jacksonville.craigslist.org   5200   
184693           frederick     https://frederick.craigslist.org   4288   

          year   manufacturer                      model  condition  \
159959  2006.0       chrysler           town country lwb       good   
122040  2013.0  mercedes-benz                  benz c250  excellent   
171067  2011.0      chevrolet             silverado 1500       good   
341640  2018.0           ford                        NaN        NaN   
413583  2019.0           ford                      f-250        NaN   
139495  2010.0           ford  f150 super cab xlt pickup       good   
265083  2016.0            bmw      5 series gran turismo  excellent   
156774  2003.0        mercury              grand marquis  excellent   
98204   2008.0         toyota                     sienna  excellent   
184693  2005.0         toyota                   camry le       good   

          cylinders  ...       size       type paint_color  \
159959          NaN  ...  full-size        van      silver   
122040  4 cylinders  ...   mid-size      sedan       white   
171067          NaN  ...  full-size      truck       white   
341640          NaN  ...        NaN        NaN         NaN   
413583  8 cylinders  ...        NaN      truck         NaN   
139495  8 cylinders  ...        NaN     pickup       white   
265083  6 cylinders  ...        NaN  hatchback        grey   
156774  8 cylinders  ...  full-size        NaN       brown   
98204   6 cylinders  ...  full-size   mini-van       white   
184693  4 cylinders  ...  full-size      sedan      silver   

                                                image_url  \
159959  https://images.craigslist.org/00v0v_3Z3NHtAIrg...   
122040  https://images.craigslist.org/00G0G_9UNdT9T9aC...   
171067  https://images.craigslist.org/00Z0Z_37ZCL1Areh...   
341640  https://images.craigslist.org/00X0X_ap99EREe8L...   
413583  https://images.craigslist.org/00z0z_3x186Tbh2V...   
139495  https://images.craigslist.org/00J0J_fImowCiSi6...   
265083  https://images.craigslist.org/00J0J_cynU5n7cKI...   
156774  https://images.craigslist.org/00C0C_iFSJiyCqJy...   
98204   https://images.craigslist.org/00000_4BpUrBo4Ae...   
184693  https://images.craigslist.org/00R0R_1jVT9OuAGu...   

                                              description county state  \
159959  2006 Chrysler Town & Country LWB Mobility Scoo...    NaN    ia   
122040  2013 Mercedes Benz C250 1.8L, V4, 4D Luxury se...    NaN    ga   
171067  100% Guaranteed Credit Approva

La commande ci-dessous, nous me permet d'avoir des informations sur le dataset comme le type de données et le nombre total des valeurs non nulles dans chaque colonne.

Cela nous permet d'identifier les valeurs manquantes et les types de données incorrectes 

In [225]:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426880 entries, 0 to 426879
Data columns (total 26 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            426880 non-null  int64  
 1   url           426880 non-null  object 
 2   region        426880 non-null  object 
 3   region_url    426880 non-null  object 
 4   price         426880 non-null  int64  
 5   year          425675 non-null  float64
 6   manufacturer  409234 non-null  object 
 7   model         421603 non-null  object 
 8   condition     252776 non-null  object 
 9   cylinders     249202 non-null  object 
 10  fuel          423867 non-null  object 
 11  odometer      422480 non-null  float64
 12  title_status  418638 non-null  object 
 13  transmission  424324 non-null  object 
 14  VIN           265838 non-null  object 
 15  drive         296313 non-null  object 
 16  size          120519 non-null  object 
 17  type          334022 non-null  object 
 18  pain

In [226]:
#Statistiques descriptives pour les colonnes numériques (moyenne, écart-type, min, max, quartiles)
df.describe().style.format('{:,.0f}') # pas de décimales ou séparateur milliers

In [227]:
# Statistiques descriptives pour des colonnes catégorielles (nombre de valeurs uniques, fréquence des plus courantes).
df.describe(include='object')

url    region  \
count                                              426880    426880   
unique                                             426880       404   
top     https://prescott.craigslist.org/cto/d/prescott...  columbus   
freq                                                    1      3608   

                            region_url manufacturer   model condition  \
count                           426880       409234  421603    252776   
unique                             413           42   29667         6   
top     https://spokane.craigslist.org         ford   f-150      good   
freq                              2988        70985    8009    121456   

          cylinders    fuel title_status transmission                VIN  \
count        249202  423867       418638       424324             265838   
unique            8       5            6            3             118264   
top     6 cylinders     gas        clean    automatic  1FMJU1JT1HEA52352   
freq          94169  356209       405117       336524                261   

         drive       size    type paint_color  \
count   296313     120519  334022      296677   
unique       3          4      13          12   
top        4wd  full-size   sedan       white   
freq    131904      63465   87056       79285   

                                                image_url  \
count                                              426812   
unique                                             241899   
top     https://images.craigslist.org/00N0N_1xMPvfxRAI...   
freq                                                 7357   

                                              description   state  \
count                                              426810  426880   
unique                                             360911      51   
top     35 VEHICLES PRICED UNDER $3000!!!  BIG TIME! T...      ca   
freq                                                  231   50614   

                    posting_date  
count                     426812  
unique                    381536  
top     2021-04-23T22:13:05-0400  
freq                          12

## Nettoyage

Cette partie consiste à manipuler le dataset dans le but de le nettoyer et de le traiter.

Pour ce faire, nous allons procéder en gérant les valeurs manquantes, doublons, abbérantes et corriger si nécessaire les types de données des colonnes non conformes



### Correction des types de données

#### Changement de posting_date en format date

Nous avons constaté que la variable posting_date est au format object, alors que nous souhaitons qu’elle soit au format date.

In [228]:
# Conversion de la colonne 'posting_date' en format datetime sans les heures
df['posting_date'] = pd.to_datetime(df['posting_date'], errors='coerce', utc=True)
df['posting_date'] = df['posting_date'].dt.tz_convert(None).dt.normalize()

In [229]:
df.sample(n=5)

id                                                url  \
294325  7305630333  https://cleveland.craigslist.org/ctd/d/gilbert...   
6941    7307561746  https://anchorage.craigslist.org/ctd/d/anchora...   
16877   7315126865  https://yuma.craigslist.org/cto/d/yuma-2004-ni...   
312599  7311181945  https://bend.craigslist.org/cto/d/bend-2005-ch...   
162691  7309901119  https://omaha.craigslist.org/ctd/d/gretna-2013...   

                        region                        region_url  price  \
294325               cleveland  https://cleveland.craigslist.org  19900   
6941        anchorage / mat-su  https://anchorage.craigslist.org   9595   
16877                     yuma       https://yuma.craigslist.org   9950   
312599                    bend       https://bend.craigslist.org   1350   
162691  omaha / council bluffs      https://omaha.craigslist.org  28988   

          year manufacturer                model condition    cylinders  ...  \
294325  2007.0         ford              f450 xl       NaN          NaN  ...   
6941    2013.0         ford               escape       NaN  4 cylinders  ...   
16877   2004.0       nissan              ud 1500      good  4 cylinders  ...   
312599  2005.0     chrysler  sebring convertible      fair  4 cylinders  ...   
162691  2013.0         ford                f-150       NaN          NaN  ...   

             size         type paint_color  \
294325        NaN          NaN         NaN   
6941          NaN          SUV      silver   
16877   full-size        truck       white   
312599        NaN  convertible       black   
162691        NaN          NaN         NaN   

                                                image_url  \
294325  https://images.craigslist.org/00q0q_c0r8yj34GW...   
6941    https://images.craigslist.org/00J0J_hhClR7HS9K...   
16877   https://images.craigslist.org/00B0B_fZHqgRynAP...   
312599  https://images.craigslist.org/01313_2oDFPerIGR...   
162691  https://images.craigslist.org/00M0M_c4Zxu7nGl9...   

                                              description county state  \
294325  **FOR SALE**  2007 Ford F450 XL Super Duty - V...    NaN    oh   
6941    This Ford Escape can be yours today! Warranty ...    NaN    ak   
16877   2004 Nissan UD 1400 truck, selling for $9950 o...    NaN    az   
312599  Selling my much-loved convertible, hoping it f...    NaN    or   
162691  Free Delivery Within 400 MilesNO REASONABLE OF...    NaN    ia   

              lat        long posting_date  
294325  42.118420  -88.435066   2021-04-12  
6941    61.154900 -149.886600   2021-04-16  
16877   32.726997 -114.626312   2021-05-01  
312599  44.151174 -121.330261   2021-04-23  
162691  41.134500  -96.245800   2021-04-20  

[5 rows x 26 columns]

#### Nettoyage de la colonne Cylinders

In [230]:
df['cylinders'].unique()

array([nan, '8 cylinders', '6 cylinders', '4 cylinders', '5 cylinders',
       'other', '3 cylinders', '10 cylinders', '12 cylinders'],
      dtype=object)

Nous retirons le mot 'cylinders' à la fin des valeurs afin de ne conserver que l’information numérique, ce qui facilitera les traitements et analyses ultérieures.

In [231]:
df['cylinders'] = df['cylinders'].astype(str).str.replace(r' cylinders', '', regex=True)

In [232]:
df['cylinders'].unique()

array(['nan', '8', '6', '4', '5', 'other', '3', '10', '12'], dtype=object)

Nous avons constaté que de nombreuses valeurs de la colonne `cylinders` sont encodées sous forme de texte `"nan"`, ce qui empêche leur détection comme valeurs manquantes réelles (`NaN`).  
Afin de garantir une imputation correcte et des statistiques fiables, nous avons standardisé cette colonne en remplaçant explicitement les chaînes vides ou équivalentes à `"nan"` par de véritables `NaN`.

In [233]:
print(df['cylinders'].isna().sum()) 

0


In [234]:
df['cylinders'] = df['cylinders'].replace(['nan'], np.nan)
print(df['cylinders'].isna().sum()) 

177678


### Gestion des valeurs manquantes

#### Imputation des variables `title_status` et `condition`

On commence par examiner les deux variables qui décrivent l’état du véhicule :

In [235]:
print(df['title_status'].unique())
print(df['condition'].unique())

[nan 'clean' 'rebuilt' 'lien' 'salvage' 'missing' 'parts only']
[nan 'good' 'excellent' 'fair' 'like new' 'new' 'salvage']


Ces deux colonnes décrivent toutes deux l’état du véhicule, mais title_status est plus objective que condition.

On examine comment ces deux variables se recoupent grâce à un tableau croisé (matrice de contingence) :

In [236]:
contingency_table = pd.crosstab(df['title_status'], df['condition'])
print(contingency_table)

condition     excellent  fair    good  like new   new  salvage
title_status                                                  
clean             91734  6156  118461     19870  1226      225
lien                586    47     389       265    25        3
missing              57   188     208         8     5       49
parts only           18    27      29         9     4       41
rebuilt            2823   142    1362       791    30       23
salvage             880   209     893       235    15      260


La majorité des véhicules avec un title_status = clean sont en good ou excellent état.

À l’inverse, title_status = salvage ou parts only est souvent associé à une condition dégradée.

On peut donc utiliser la valeur de title_status pour estimer raisonnablement les valeurs manquantes dans condition.

In [237]:
# Remplacer les NaN de title_status par 'missing'
df['title_status'] = df['title_status'].fillna('missing')

# Certaines valeurs de title_status laissent peu de doute sur la condition
# On applique une imputation basée sur la logique métier : un véhicule "salvage" est sûrement aussi "salvage" en condition
mapping_condition = {
    'salvage': 'salvage',
    'parts only': 'salvage'
}

for status, cond in mapping_condition.items():
    df.loc[df['condition'].isna() & (df['title_status'] == status), 'condition'] = cond
    
# Pour les cas restants, on choisit d’imputer les valeurs manquantes de condition par la valeur la plus fréquente pour chaque statut
def impute_with_mode(group):
    mode_val = group.mode()
    return group.fillna(mode_val[0] if not mode_val.empty else 'unknown condition')

df['condition'] = df.groupby('title_status')['condition'].transform(impute_with_mode)

# On vérifie qu’il ne reste aucune valeur manquante dans condition, et que la distribution est cohérente.
print(df['condition'].isna().sum())  # Doit renvoyer 0
print(pd.crosstab(df['title_status'], df['condition']))

0
condition     excellent  fair    good  like new   new  salvage
title_status                                                  
clean             91734  6156  285906     19870  1226      225
lien                693    47     389       265    25        3
missing            8484   188     322         8     5       49
parts only           18    27      29         9     4      111
rebuilt            4871   142    1362       791    30       23
salvage             880   209     893       235    15     1636


En conclusion, nous avons mené une exploration approfondie des variables `title_status` et `condition`, qui décrivent toutes deux l’état du véhicule. L’analyse a révélé une forte corrélation entre ces deux colonnes, avec une information plus fiable et objective dans `title_status`. Cette observation nous a permis d’adopter une stratégie d’imputation conditionnelle cohérente pour compléter les valeurs manquantes dans `condition`. On a d’abord appliqué une logique métier simple pour certains statuts très parlants comme "salvage" ou "parts only", puis complété les valeurs restantes en utilisant la modalité la plus fréquente pour chaque statut. Cette démarche a permis de restaurer la complétude de la variable `condition` tout en conservant une cohérence interne dans les données, ce qui améliore la qualité globale du jeu de données et garantit des analyses plus fiables par la suite.

#### Imputation des valeurs manquantes dans les variables `manufacturer`, `drive`, `cylinders`, `fuel`, `transmission`, `size` et `type`

Afin de compléter les valeurs manquantes dans certaines colonnes du jeu de données, nous avons mis en place une fonction d’imputation appliquée à chaque combinaison unique de model et d'une colonne cible (par exemple : drive, cylinders, fuel, transmission…). Cette fonction suit une logique en quatre étapes :

Aucun NaN présent dans la série : la série est retournée telle quelle, sans modification.

Toutes les valeurs sont manquantes (que des NaN) : la série est également retournée sans modification, car aucune information ne permet une imputation fiable.

Une seule valeur observée (non nulle) : tous les NaN sont remplacés par cette valeur unique, considérée comme représentative du modèle.

Plusieurs valeurs observées : les NaN sont imputés aléatoirement, en respectant la distribution empirique des valeurs existantes. Cela permet de conserver la variabilité des données tout en remplissant les valeurs manquantes de manière cohérente.

Cette approche permet une imputation contextualisée et réaliste, en s'appuyant sur la cohérence interne des données pour chaque modèle.

In [238]:
# Dans un premier temps, la liste cols identifie les colonnes sur lesquelles l’imputation va porter : 'manufacturer', 'drive', 'cylinders', 'fuel' et 'transmission'
cols = ['manufacturer', 'drive', 'cylinders', 'fuel', 'transmission', 'size' , 'type']

# Une copie du DataFrame original est réalisée sous le nom df_pre, afin de conserver l’état initial des données avant modification
df_pre = df.copy()

# On calcule et affiche le nombre de valeurs manquantes (NaN) pour chacune de ces colonnes
na_counts = df[cols].isna().sum()
print("NaN par colonne :\n", na_counts, "\n")

# Pour comprendre comment procéder à l’imputation, le code examine combien de valeurs différentes existent pour chaque variable (cols) à l’intérieur de chaque modèle de voiture (colonne model). 
# Cela permet d’identifier les cas où tous les enregistrements d’un même modèle partagent une seule et même valeur (cas simple) et ceux où plusieurs valeurs coexistent (cas complexe). 
# Le résultat est résumé dans un DataFrame summary.
uniques = df.groupby('model')[cols].nunique()
summary = pd.DataFrame({
    'single_value_models':  (uniques == 1).sum(),
    'multiple_value_models': (uniques > 1).sum()
})
print("Nombre de modèles par colonne → 1 seule valeur vs plusieurs :\n", summary, "\n")

# Un générateur de nombres aléatoires, initialisé avec une graine fixe (42), est ensuite défini avec np.random.default_rng(42). 
# Cela garantit que les tirages aléatoires réalisés plus tard seront reproductibles
rng = np.random.default_rng(42)

# On calcule pour chaque colonne concernée, le mode global, c’est-à-dire la valeur la plus fréquente dans l’ensemble du DataFrame (hors NaN). Ces modes globaux serviront de solution de secours si aucune information n’est disponible localement.
global_modes = {col: df[col].mode(dropna=True)[0] for col in cols}

# La fonction impute_group(x, col) définit la stratégie d’imputation pour un sous-ensemble de données x, correspondant à une colonne col dans un modèle de voiture donné
def impute_group(x, col):
    """
    Impute les NaN dans une série x (un seul model pour la colonne col).
    - Si aucun NaN           : renvoie x tel quel.
    - Si tout est NaN         : remplit par le mode global.
    - Si une seule valeur    : duplique cette valeur pour tous les NaN.
    - Sinon                   : tirage aléatoire pondéré par la fréquence locale.
    """
    non_na = x.dropna()
    n_miss = x.isna().sum()

    # Si la série ne contient aucune valeur manquante, elle est retournée telle quelle
    if n_miss == 0:
        return x

    if non_na.empty:
        # fallback global si aucun échantillon local
        return x.fillna(global_modes[col])
    # Si une seule valeur non nulle existe, elle est utilisée pour remplacer les NaN
    if non_na.nunique() == 1:
        return x.fillna(non_na.iloc[0])

    # Sinon, on effectue un tirage aléatoire pondéré parmi les valeurs existantes, selon leur fréquence.
    dist = non_na.value_counts(normalize=True)
    fill_vals = rng.choice(dist.index, size=n_miss, p=dist.values)
    x.loc[x.isna()] = fill_vals
    return x

# --- Imputation ---
# Le cœur de l’imputation se fait dans une boucle for sur chaque colonne de la liste cols
for col in cols:
    # Pour chaque colonne, on applique la fonction impute_group groupe par groupe, en fonction de la variable model. Cela constitue l’imputation locale.
    filled = (
        df
        .groupby('model')[col]
        .transform(lambda s: impute_group(s, col))
    )
    # Une fois cette opération réalisée, les valeurs restant éventuellement manquantes (si aucune valeur valide n'était présente dans un groupe) sont remplacées par le mode global.
    filled = filled.fillna(global_modes[col])
    # Enfin, le résultat est affecté à la colonne correspondante du DataFrame.
    df[col] = filled

# À la fin du processus, un dernier affichage vérifie qu’il ne reste plus aucune valeur manquante dans les colonnes traitées
print("NaN restants :", df[cols].isna().sum().to_dict())

# Le DataFrame final, après imputation, est copié sous le nom df_post
df_post = df.copy()

NaN par colonne :
 manufacturer     17646
drive           130567
cylinders       177678
fuel              3013
transmission      2556
size            306361
type             92858
dtype: int64 

Nombre de modèles par colonne → 1 seule valeur vs plusieurs :
               single_value_models  multiple_value_models
manufacturer                23481                    272
drive                       20679                   2491
cylinders                   19885                   2352
fuel                        25901                   3091
transmission                26093                   3504
size                        10408                   2833
type                        20613                   3944 

NaN restants : {'manufacturer': 0, 'drive': 0, 'cylinders': 0, 'fuel': 0, 'transmission': 0, 'size': 0, 'type': 0}


Pour vérifier que les propotions sont bien respectées, nous allons d’abord repéré pour chaque variable (manufacturer, drive, cylinders, fuel et transmission) les modèles dont le taux de valeurs manquantes, avant traitement, se situait strictement entre 5 % et 100 %, puis, parmi ces modèles, sélectionner les trois présentant le plus grand nombre de modalités distinctes. Pour chacun de ces modèles, il va calculé la répartition proportionnelle des valeurs avant imputation, puis la même répartition après imputation, en arrondissant chaque proportion à deux décimales. La présentation côte-à-côte des distributions avant et après traitement nous a permis de vérifier que les proportions relatives des différentes modalités étaient conservées et qu’aucun biais notable n’avait été introduit par le processus de remplissage des données manquantes.

In [239]:
cols = ['manufacturer', 'drive', 'cylinders', 'fuel', 'transmission', 'size' , 'type']

for col in cols:
    # a) Taux de NaN et nombre de valeurs distinctes par model AVANT
    grp = df_pre.groupby('model')[col]
    nan_rate = grp.apply(lambda s: s.isna().mean())
    nunique = grp.nunique()

    # b) Filtrer modèles 5% < NaN < 100%
    valid = nunique.index[(nan_rate > 0.05) & (nan_rate < 1.0)]

    # c) Top 3 modèles selon nunique pour cette colonne
    top3 = nunique.loc[valid].nlargest(3).index.tolist()
    print(f"\n=== Colonne `{col}` — Top 3 modèles ===\n{top3}")

    # d) Afficher pour chacun la distribution avant/après, arrondie à 2 décimales
    for model in top3:
        dist_pre = (
            df_pre.loc[df_pre['model'] == model, col]
                  .value_counts(normalize=True)
        )
        dist_post = (
            df_post.loc[df_post['model'] == model, col]
                   .value_counts(normalize=True)
        )
        comp = pd.DataFrame({'avant': dist_pre, 'après': dist_post}) \
                 .fillna(0) \
                 .sort_index() \
                 .round(2)

        print(f"\nModèle `{model}` (NaN avant : {nan_rate[model]:.2f}, valeurs distinctes : {nunique[model]})")
        print(comp)


=== Colonne `manufacturer` — Top 3 modèles ===
['4x4', 'cargo van', 'custom']

Modèle `4x4` (NaN avant : 0.08, valeurs distinctes : 6)
                 avant  après
manufacturer                 
chevrolet         0.17   0.15
dodge             0.08   0.08
harley-davidson   0.08   0.08
jeep              0.08   0.08
ram               0.25   0.23
toyota            0.33   0.38

Modèle `cargo van` (NaN avant : 0.05, valeurs distinctes : 5)
              avant  après
manufacturer              
chevrolet      0.20   0.19
dodge          0.03   0.03
ford           0.09   0.08
gmc            0.09   0.11
ram            0.60   0.59

Modèle `custom` (NaN avant : 0.06, valeurs distinctes : 5)
              avant  après
manufacturer              
chevrolet      0.07   0.06
fiat           0.07   0.06
ford           0.67   0.69
gmc            0.10   0.09
mercury        0.10   0.09

=== Colonne `drive` — Top 3 modèles ===
['-', '1500', '1500 crew cab']

Modèle `-` (NaN avant : 0.75, valeurs distinctes :

### Suppresion des lignes inutiles ou avec des valeurs manquantes importantes

##### Nettoyage des colonnes non pertinentes

Dans le cadre de notre analyse, nous avons décidé de supprimer plusieurs colonnes du jeu de données initial afin de le simplifier et de conserver uniquement les variables utiles à notre étude. Ce choix repose sur les critères suivants :

- **`id`** : Non pertinent pour l’analyse, car il s’agit uniquement d’un identifiant technique.
- **`region`** : Redondant avec la variable state, qui suffira pour localiser les annonces.
- **`url`**, **`region_url`**, **`image_url`** : ces colonnes contiennent uniquement des liens externes qui ne présentent aucune utilité analytique directe. Elles ne contribuent pas à la compréhension du marché de la voiture d’occasion.
- **`description`** : bien que cette colonne contienne potentiellement des informations intéressantes, elle n’est pas exploitable telle quelle sans un traitement par des techniques de traitement automatique du langage naturel (NLP), ce qui dépasse le périmètre de notre analyse.
- **`paint_color`** : cette variable est très incomplète dans le jeu de données. Son taux de remplissage est trop faible pour être exploité de manière fiable.
- **`county`** : cette colonne est presque entièrement vide. Elle n’apporte donc aucune information exploitable.
- **`lat`** et **`long`** : les coordonnées géographiques précises ne sont pas nécessaires pour notre analyse, puisque nous disposons déjà de la variable **`state`**, suffisante pour les regroupements régionaux.

En résumé, ces colonnes ont été jugées non pertinentes, soit parce qu'elles sont inutilisables dans l’état actuel, soit parce qu’elles ne contribuent pas à notre objectif d’analyse.

In [240]:
df = df.drop(columns=['id', 'region', 'url', 'region_url', 'image_url', 'description', 'paint_color', 'county', 'lat', 'long'])

##### Suppression des lignes avec des valeurs "parts only" et "salvage"

Pour se concentrer sur les véhicules d’occasion susceptibles d’être vendus à des particuliers dans un état de fonctionnement, on a choisi d’exclure les statuts "parts only" et "salvage", qui correspondent à des voitures non roulantes ou accidentées.

In [241]:
# Nombre de lignes avant filtrage
nb_avant = len(df)

df = df[~df['title_status'].isin(['parts only', 'salvage'])]

# Nombre de lignes après filtrage
nb_apres = len(df)

# Affichage
print(f"Nombre de lignes avant filtrage : {nb_avant}")
print(f"Nombre de lignes après filtrage : {nb_apres}")
print(f"Nombre de lignes supprimées : {nb_avant - nb_apres} ({((nb_avant-nb_apres)/nb_avant)*100:.2f}%)")


Nombre de lignes avant filtrage : 426880
Nombre de lignes après filtrage : 422814
Nombre de lignes supprimées : 4066 (0.95%)


#### Suppression des lignes avec des valeurs manquantes critiques

Certaines colonnes sont essentielles pour notre analyse du marché des véhicules d’occasion. Nous avons identifié les colonnes suivantes comme étant indispensables :

- **`price`** : le prix est la variable cible principale de notre étude.
- **`year`** : l’année du véhicule permet d’analyser la dépréciation et le vieillissement.
- **`manufacturer`** et **`model`** : ces deux variables définissent précisément le véhicule concerné.
- **`odometer`** : le kilométrage est un critère fondamental pour estimer la valeur d’un véhicule.

Les lignes contenant des valeurs manquantes dans l’une de ces colonnes ont donc été supprimées, car elles ne pourraient pas être utilisées de manière fiable dans les analyses à venir.

In [242]:
# Nombre de lignes avant filtrage
nb_avant = len(df)

colonnes_importantes = ['price', 'year', 'manufacturer', 'model', 'odometer']
df = df.dropna(subset=colonnes_importantes)

# Nombre de lignes après filtrage
nb_apres = len(df)

# Affichage
print(f"Nombre de lignes avant filtrage : {nb_avant}")
print(f"Nombre de lignes après filtrage : {nb_apres}")
print(f"Nombre de lignes supprimées : {nb_avant - nb_apres} ({((nb_avant-nb_apres)/nb_avant)*100:.2f}%)")

Nombre de lignes avant filtrage : 422814
Nombre de lignes après filtrage : 412174
Nombre de lignes supprimées : 10640 (2.52%)


### Gestion des doublons

#### Premier dédoublonnage en utilisant l'immatriculation du véhicule `VIN` et la date du posting `posting_date`

Ce traitement vise à supprimer les doublons présents dans le DataFrame `df`, mais uniquement pour les lignes contenant un **numéro VIN** (*Vehicle Identification Number*). Le numéro VIN permet d’identifier un véhicule de manière unique, ce qui est essentiel pour détecter les annonces en doublon.

Étape 1 : Isolation des lignes avec un VIN valide

On commence par créer un sous-ensemble `df_vin_non_null`, qui contient uniquement les lignes où la colonne `VIN` est renseignée (c’est-à-dire différente de vide (`''`) et de valeur manquante (`NaN`)). Ces lignes représentent les véhicules identifiables de manière unique.

Étape 2 : Suppression des doublons sur VIN et date de publication

Une fois ce sous-ensemble isolé, on applique la méthode `drop_duplicates()` en se basant sur les colonnes `VIN` et `posting_date`. Cela permet de supprimer les annonces répétées pour un même véhicule publiées à la même date. Le résultat est stocké dans le DataFrame `df_vin_dedoublonne`.

Étape 3 : Conservation des lignes sans VIN

En parallèle, on extrait les lignes qui **ne possèdent pas de VIN valide**. Ces entrées sont conservées dans un second sous-ensemble `df_sans_vin`, car bien qu’elles ne puissent pas être dédupliquées sur la base du VIN, elles peuvent rester utiles pour certaines analyses.

Étape 4 : Fusion des deux sous-ensembles

Enfin, on combine les deux DataFrames (`df_vin_dedoublonne` et `df_sans_vin`) en un seul DataFrame final grâce à `pd.concat()`, avec l’option `ignore_index=True` pour réinitialiser les index. Le DataFrame `df` contient ainsi toutes les données originales, avec les doublons supprimés uniquement là où cela était possible et pertinent.

In [243]:
# Exclure les lignes sans VIN (on ne veut pas les traiter comme doublons)
# On recupère les voitures avec Matricules
df_vin_non_null = df[df['VIN'].notna() & (df['VIN'] != '')]

# Dédoublonnage sur (VIN + posting_date) parmi les lignes valides
df_vin_dedoublonne = df_vin_non_null.drop_duplicates(subset=['VIN', 'posting_date'])

# Garder les autres lignes non concernées (avec VIN vide ou NaN)
df_sans_vin = df[df['VIN'].isna() | (df['VIN'] == '')]

# Recombiner les deux
df = pd.concat([df_vin_dedoublonne, df_sans_vin], ignore_index=True)

# === Résumé du dédoublonnage ===
n_total_avant = df_vin_non_null.shape[0]
n_total_apres = df_vin_dedoublonne.shape[0]
n_doublons = n_total_avant - n_total_apres
pct_doublons = (n_doublons / n_total_avant) * 100
n_sans_vin = df_sans_vin.shape[0]
n_total_final = df.shape[0]

print("\n=== Résumé du dédoublonnage ===")
print(f"Total initial de lignes avec VIN : {n_total_avant}")
print(f"Nombre de doublons écartés (même VIN + posting_date) : {n_doublons} ({pct_doublons:.2f}%)")
print(f"Total de lignes conservées avec VIN : {n_total_apres}")
print(f"Total de lignes sans VIN (non traitées) : {n_sans_vin}")
print(f"Total final après recombinaison : {n_total_final}")

# Supprimée VIN après le dédoublonnage, car elle n'est plus nécessaire pour l’analyse.
df = df.drop(columns='VIN')


=== Résumé du dédoublonnage ===
Total initial de lignes avec VIN : 259463
Nombre de doublons écartés (même VIN + posting_date) : 85892 (33.10%)
Total de lignes conservées avec VIN : 173571
Total de lignes sans VIN (non traitées) : 152711
Total final après recombinaison : 326282


#### Deuxième dédoublonnage sur les lignes sans 'VIN' en utilisant les variables importantes et la date du posting 'posting_date'

Après avoir effectué un premier nettoyage basé sur le numéro VIN, nous réalisons un second dédoublonnage plus général, basé cette fois sur des **attributs clés** qui décrivent le véhicule de manière précise.

L’objectif est de repérer et supprimer les doublons pour lesquels le véhicule présente exactement les mêmes valeurs sur plusieurs colonnes caractéristiques. Ce type de doublon peut apparaître lorsqu'une même annonce est saisie plusieurs fois avec des identifiants manquants ou différents, mais des attributs identiques.

Nous avons retenu les colonnes suivantes comme représentatives d’un véhicule unique dans le contexte d’une annonce :

- **`price`** : le prix de vente proposé
- **`year`** : l’année du véhicule
- **`manufacturer`** : le constructeur
- **`model`** : le modèle
- **`posting_date`** : la date de publication de l’annonce

La méthode `drop_duplicates()` est appliquée en se basant sur ces colonnes. Seules les lignes présentant une combinaison strictement identique de ces valeurs sont considérées comme des doublons et supprimées.
Ce dédoublonnage complémentaire permet d’assurer une meilleure qualité des données en éliminant les répétitions involontaires qui ne peuvent pas être détectées via le numéro VIN.


In [244]:
# Définir les colonnes clés pour le second dédoublonnage
colonnes_importantes = ['price', 'year', 'manufacturer', 'model', 'posting_date']
n_avant = df.shape[0]

# Appliquer le drop_duplicates
df = df.drop_duplicates(subset=colonnes_importantes)
n_apres = df.shape[0]

# === Résumé du dédoublonnage ===
n_doublons = n_avant - n_apres
pct_doublons = (n_doublons / n_avant) * 100

print("\n=== Résumé du dédoublonnage (par attributs clés) ===")
print(f"Total initial de lignes : {n_avant}")
print(f"Nombre de doublons écartés (même combinaison colonnes importantes) : {n_doublons} ({pct_doublons:.2f}%)")
print(f"Total final après dédoublonnage : {n_apres}")


=== Résumé du dédoublonnage (par attributs clés) ===
Total initial de lignes : 326282
Nombre de doublons écartés (même combinaison colonnes importantes) : 28611 (8.77%)
Total final après dédoublonnage : 297671


### Traitement des valeurs abérrantes (outlet)

#### Gestion des valeurs aberrantes – Variable `odometer`

On utilise ce code pour analyser la répartition des valeurs extrêmes de odometer, afin d’identifier des seuils pertinents pour le nettoyage des données avant toute analyse.

In [245]:
# Total de lignes pour référence
total_lignes = df.shape[0]

# Seuils inférieurs
seuils_inferieurs = [0, 10, 50, 100, 500, 1000]
print("=== Nombre de lignes avec odometer inférieur ou égal aux seuils ===")
for seuil in seuils_inferieurs:
    count = (df['odometer'] <= seuil).sum()
    print(f"  ≤ {seuil} miles : {count} lignes ({(count/total_lignes)*100:.2f}%)")

# Seuils supérieurs
seuils_superieurs = [150000, 200000, 300000, 400000, 500000]
print("\n=== Nombre de lignes avec odometer supérieur aux seuils ===")
for seuil in seuils_superieurs:
    count = (df['odometer'] > seuil).sum()
    print(f"  > {seuil} miles : {count} lignes ({(count/total_lignes)*100:.2f}%)")



=== Nombre de lignes avec odometer inférieur ou égal aux seuils ===
  ≤ 0 miles : 1159 lignes (0.39%)
  ≤ 10 miles : 3546 lignes (1.19%)
  ≤ 50 miles : 4039 lignes (1.36%)
  ≤ 100 miles : 4505 lignes (1.51%)
  ≤ 500 miles : 6136 lignes (2.06%)
  ≤ 1000 miles : 7094 lignes (2.38%)

=== Nombre de lignes avec odometer supérieur aux seuils ===
  > 150000 miles : 62858 lignes (21.12%)
  > 200000 miles : 19271 lignes (6.47%)
  > 300000 miles : 2399 lignes (0.81%)
  > 400000 miles : 1274 lignes (0.43%)
  > 500000 miles : 1046 lignes (0.35%)


Après analyse de la distribution des valeurs et en croisant les recommandations issues de sources métier, nous avons choisi de restreindre la variable `odometer` aux valeurs comprises entre 100 miles et 300 000 miles**.

Justification du seuil bas (100 miles)

Les véhicules affichant un kilométrage très faible sont probablement mal renseignés ou mal catégorisés (véhicules neufs enregistrés comme voitures d'occasion, erreurs de saisie, etc.).

Dans le dataset, on observe :
- 1 209 lignes avec un kilométrage ≤ 10 miles  
- 3 719 lignes ≤ 50 miles  
- 4 292 lignes ≤ 100 miles

Ces lignes représentent **moins de 1,5 % de l’ensemble du dataset** (371 031 lignes). Leur suppression permet de réduire le bruit, d’éviter des valeurs aberrantes et d’améliorer la lisibilité des visualisations, sans compromettre la qualité de l’échantillon analysé.

Justification du seuil haut (300 000 miles)

Selon la source métier [Mercier Auto]( https://www.mercier-auto.com/guide-achat-auto/combien-de-kilometrage-pour-une-voiture-doccasion/#:~:text=Il%20peut%20s'av%C3%A9rer%20difficile,s'%C3%A9l%C3%A8ve%20%C3%A0%20150.000%20kilom%C3%A8tres.), la revente devient difficile à partir de :
- 150 000 km (soit environ 93 000 miles) pour les véhicules essence,
- 150 000 miles pour les véhicules diesel.

Si on appliquait ces seuils strictement, on exclurait :
- 64 854 lignes pour les véhicules entre 150 000 et 200 000 miles 
- 20 086 lignes supplémentaires entre 200 000 et 300 000 miles

Cela représenterait **près de 23 % du dataset**, ce qui est **excessif** par rapport à l’objectif initial de nettoyage.  
En retenant une limite haute à **300 000 miles**, on élimine les valeurs les plus extrêmes et les cas potentiellement erronés, tout en conservant les données réalistes pour des usages professionnels (taxis, utilitaires, flottes longue distance...).

Conclusion

Le choix de filtrer `odometer` entre **100 et 300 000 miles** repose donc sur trois critères :
- Une **analyse quantitative** des données disponibles
- Une **référence métier** crédible sur les seuils de revente réalistes
- Une **volonté de préserver la lisibilité des graphiques** et la représentativité de l’échantillon

Ce nettoyage ciblé permet d’améliorer la qualité globale du dataset sans perte significative d’information.


In [246]:
# Nombre de lignes avant filtrage
nb_avant = len(df)

#Suppression des kilométrages incohérents
df = df[(df['odometer'] >= 100) & (df['odometer'] <= 300000)]

# Nombre de lignes après filtrage
nb_apres = len(df)

# Affichage
print(f"Nombre de lignes avant filtrage : {nb_avant}")
print(f"Nombre de lignes après filtrage : {nb_apres}")
print(f"Nombre de lignes supprimées : {nb_avant - nb_apres} ({((nb_avant-nb_apres)/total_lignes)*100:.2f}%)")

Nombre de lignes avant filtrage : 297671
Nombre de lignes après filtrage : 290980
Nombre de lignes supprimées : 6691 (2.25%)


#### Gestion des valeurs aberrantes – Variable `year`

On a analysé la variable `year` afin d’identifier et exclure les années aberrantes avant de poursuivre l’analyse. Étant donné que les dates de publication (`posting_date`) dans le dataset s’arrêtent à 2021, cette année a été retenue comme **seuil maximum** pour les véhicules.

Concernant les années les plus anciennes, on a compté le nombre de lignes correspondant à différentes bornes. Les résultats sont les suivants :

In [247]:
# Examiner les années extrêmes pour définir des seuils
print("Année min :", df['year'].min())
print("Année max :", df['year'].max())

Année min : 1900.0
Année max : 2022.0


In [248]:
# Total de lignes pour référence
total_lignes = df.shape[0]

# Seuils pour années aberrantes
seuils_bas = [1900, 1950, 1960, 1970, 1980]

print("=== Lignes avec year ≤ seuils ===")
for seuil in seuils_bas:
    count = (df['year'] <= seuil).sum()
    print(f"  ≤ {seuil} : {count} lignes ({(count/total_lignes)*100:.2f}%)")


=== Lignes avec year ≤ seuils ===
  ≤ 1900 : 11 lignes (0.00%)
  ≤ 1950 : 795 lignes (0.27%)
  ≤ 1960 : 1583 lignes (0.54%)
  ≤ 1970 : 3852 lignes (1.32%)
  ≤ 1980 : 6292 lignes (2.16%)


On observe que les véhicules antérieurs à 1950 représentent seulement **0,34 % du dataset**. Ces cas sont rares et probablement mal renseignés, ou correspondent à des véhicules anciens ou de collection peu représentatifs du marché de l’occasion tel qu’il est analysé ici.

On a donc choisi de conserver uniquement les véhicules dont l’année de fabrication est comprise entre **1950 et 2021**, ce qui permet :
- d’exclure uniquement les cas extrêmes
- de conserver l’essentiel des données valides, y compris des modèles anciens
- et d’améliorer la fiabilité des analyses à venir

In [249]:
# Nombre de lignes avant filtrage
nb_avant = len(df)

#Suppression des années incohérentes
df = df[(df['year'] >= 1950) & (df['year'] <= 2021)]

# Nombre de lignes après filtrage
nb_apres = len(df)

# Affichage
print(f"Nombre de lignes avant filtrage : {nb_avant}")
print(f"Nombre de lignes après filtrage : {nb_apres}")
print(f"Nombre de lignes supprimées : {nb_avant - nb_apres} ({((nb_avant-nb_apres)/total_lignes)*100:.2f}%)")

Nombre de lignes avant filtrage : 290980
Nombre de lignes après filtrage : 290223
Nombre de lignes supprimées : 757 (0.26%)


#### Gestion des valeurs aberrantes – Variable `price`

Avant d’appliquer tout traitement sur la variable `price`, il est important de bien comprendre sa distribution à l’aide de statistiques descriptives de base.

Nous utilisons la méthode `describe()` pour obtenir les principales mesures statistiques : minimum, maximum, moyenne, médiane (50%), écart interquartile, etc.


In [250]:
# Statistiques descriptives
print(df['price'].describe())

# Valeurs uniques les plus faibles
print("Valeurs les plus basses :", sorted(df['price'].unique())[:20])

# Valeurs les plus hautes
print("Valeurs les plus hautes :", sorted(df['price'].unique())[-20:])

count    2.902230e+05
mean     7.730073e+04
std      1.326278e+07
min      0.000000e+00
25%      5.300000e+03
50%      1.099900e+04
75%      2.179600e+04
max      3.736929e+09
Name: price, dtype: float64
Valeurs les plus basses : [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
Valeurs les plus hautes : [990000, 999999, 1000000, 1111111, 1234567, 2000000, 3226714, 6995495, 10004000, 11111111, 17000000, 25003000, 99999999, 123456789, 135008900, 1111111111, 1234567890, 1410065407, 3024942282, 3736928711]


Afin de mieux comprendre la répartition des valeurs extrêmes dans la variable `price`, nous avons défini plusieurs seuils représentatifs à la fois pour les valeurs très faibles et très élevées. L’objectif est de quantifier la proportion de lignes concernées par ces cas limites.

Nous avons établi deux séries de seuils :

- **Seuils bas** : `[0, 100, 500, 1000, 2000]`  
  Ces seuils permettent d’identifier les valeurs anormalement faibles, voire incohérentes dans le contexte d’une vente (ex. : prix = 0 ou 1).

- **Seuils hauts** : `[50000, 100000, 150000, 200000, 500000]`  
  Ils nous permettent de repérer les véhicules très haut de gamme, les erreurs de saisie ou les cas hors-norme.


In [251]:
total_lignes = df.shape[0]

seuils_bas = [0, 100, 500, 1000, 2000]
seuils_haut = [50000, 100000, 150000, 200000, 500000]

print("=== Lignes avec price ≤ seuils ===")
for seuil in seuils_bas:
    count = (df['price'] <= seuil).sum()
    print(f"  ≤ {seuil} : {count} lignes ({(count/total_lignes)*100:.2f}%)")

print("\n=== Lignes avec price ≥ seuils ===")
for seuil in seuils_haut:
    count = (df['price'] >= seuil).sum()
    print(f"  ≥ {seuil} : {count} lignes ({(count/total_lignes)*100:.2f}%)")

=== Lignes avec price ≤ seuils ===
  ≤ 0 : 19014 lignes (6.55%)
  ≤ 100 : 21071 lignes (7.26%)
  ≤ 500 : 26080 lignes (8.99%)
  ≤ 1000 : 29431 lignes (10.14%)
  ≤ 2000 : 35158 lignes (12.11%)

=== Lignes avec price ≥ seuils ===
  ≥ 50000 : 7810 lignes (2.69%)
  ≥ 100000 : 409 lignes (0.14%)
  ≥ 150000 : 120 lignes (0.04%)
  ≥ 200000 : 72 lignes (0.02%)
  ≥ 500000 : 35 lignes (0.01%)


Avant d'exploiter la variable `price`, il est essentiel d'examiner la distribution de ses valeurs extrêmes, aussi bien très faibles que très élevées, afin d’identifier les cas aberrants ou potentiellement erronés.

Nous constatons que :

- 18 920 lignes (6.51%) ont un prix **à 0**, ce qui est incohérent dans le cadre d'une vente.
- 21 003 lignes (7.22%) ont un prix **inférieur ou égal à 100**.
- 26 019 lignes (8.95%) ont un prix **inférieur ou égal à 500**.
- 29 376 lignes (10.10%) ont un prix **inférieur ou égal à 1000**.
- 35 118 lignes (12.08%) ont un prix **inférieur ou égal à 2000**.

Une part non négligeable du jeu de données présente donc des valeurs très faibles, dont certaines (notamment 0) sont probablement dues à des erreurs de saisie ou à des annonces incomplètes.

À l’inverse, on observe également des valeurs très élevées :

- 7 842 lignes (2.70%) ont un prix **supérieur ou égal à 50 000**.
- 412 lignes (0.14%) dépassent **100 000**.
- 120 lignes (0.04%) dépassent **150 000**.
- 72 lignes (0.02%) dépassent **200 000**.
- 35 lignes (0.01%) atteignent ou dépassent **500 000**.

Ces observations suggèrent la présence de véhicules de luxe ou de valeurs mal renseignées, à vérifier selon le contexte de l’annonce.

À la suite de l’analyse statistique des extrêmes, nous avons décidé de ne conserver que les annonces dont le prix est :

- **supérieur à 500** : afin d’éliminer les valeurs anormalement faibles ou nulles, souvent liées à des erreurs de saisie, des annonces incomplètes ou des réservations de visibilité.
- **inférieur à 150 000** : afin d’écarter les véhicules très haut de gamme, les erreurs d’encodage ou les valeurs atypiques qui pourraient fortement déséquilibrer les analyses statistiques.

Ce filtrage permet de nous concentrer sur la tranche de prix la plus représentative du marché des voitures d’occasion, tout en supprimant environ 14 % de valeurs extrêmes peu fiables ou non représentatives.

In [252]:
# Nombre de lignes avant filtrage
nb_avant = len(df)

#Suppression des prix incohérents
df = df[(df['price'] >= 500) & (df['price'] <= 150000)]

# Nombre de lignes après filtrage
nb_apres = len(df)

# Affichage
print(f"Nombre de lignes avant filtrage : {nb_avant}")
print(f"Nombre de lignes après filtrage : {nb_apres}")
print(f"Nombre de lignes supprimées : {nb_avant - nb_apres} ({((nb_avant-nb_apres)/total_lignes)*100:.2f}%)")

Nombre de lignes avant filtrage : 290223
Nombre de lignes après filtrage : 264657
Nombre de lignes supprimées : 25566 (8.81%)


## Axes d'analyse

1. Distribution des prix
Histogramme des prix

2. Décote du prix selon l'âge du véhicule
Variable à créer véhicle_age = posting_date - year

3. Analyse géographique par Etat
Nom d'annonce par Etat

4. Prix par kilométrage
Relation prix vs kilométrage et condition

5. Analyse par marque
Par model et filtrage caractéristique voiture

## Transformer le dataframe nettoyé en csv

In [253]:
df.to_csv('vehicles_clean.csv', index=False)